src/v/cluster/data_migration_backend.cc

```diff
@@ -40,27 +54,753 @@ backend::backend(
+      model::controller_ntp, _plt_raft0_leadership_notification_id);
+    _table.unregister_notification(_table_notification_id);
+    co_await _gate.close();
+}
+
+ss::future<> backend::loop_once() {
+    {
+        auto locked = co_await lock();
+        if (!locked) {
+            co_return;
+        }
+        co_await work_once();
+    }
+    co_await _sem.wait(_as);
+    _sem.consume(_sem.available_units());
 }
 
-void backend::handle_migration_update(id id) {
+ss::future<> backend::work_once() {
+    // process pending deltas
+    auto unprocessed_deltas = std::move(_unprocessed_deltas);
+    for (auto&& delta : unprocessed_deltas) {
+        co_await process_delta(std::move(delta));
+    }
+
+    // process RPC responses
+    for (const auto& [node_id, response] : _rpc_responses) {
+        co_await ssx::async_for_each(
+          response.actual_states, [this](const auto& ntp_resp) {
+              auto rs_it = _mrstates.find(ntp_resp.migration);
+              if (rs_it == _mrstates.end()) {
+                  // migration gone, ignore
+                  return;
+              }
+              migration_reconciliation_state& rs = rs_it->second;
+              if (rs.sought_state > ntp_resp.state) {
+                  // migration advanced since then, ignore
+                  return;
+              }
+              mark_rs_done_for_ntp(rs, ntp_resp.ntp);
+              if (rs.outstanding_topics.empty()) {
+                  to_advance(ntp_resp.migration, rs.sought_state);
+                  _mrstates.erase(rs_it);
+              }
+          });
+    }
+
+    auto alarm = model::timeout_clock::time_point::max();
+
+    // prepare RPC requests
+    chunked_vector<model::node_id> to_send_rpc;
+    auto now = model::timeout_clock::now();
+    for (const auto& [node_id, deadline] : _nodes_to_retry) {
+        if (deadline <= now) {
+            to_send_rpc.push_back(node_id);
+        } else {
+            alarm = std::min(deadline, alarm);
+        }
+    }
+
+    // defer RPC retries
+    // todo: configure timeout
+    auto new_deadline = now + 5s;
+    for (const auto& node_id : _rpc_responses | std::views::keys) {
+        if (_nstates.count(node_id)) {
+            _nodes_to_retry.try_emplace(node_id, new_deadline);
+            alarm = std::min(alarm, new_deadline);
+        }
+    }
+    _rpc_responses.clear();
+
+    // schedule fibers
+    for (auto node_id : to_send_rpc) {
+        _nodes_to_retry.erase(node_id);
+        co_await send_rpc(node_id);
+    }
+    spawn_advances();
+    if (alarm == model::timeout_clock::time_point::max()) {
+        _timer.cancel();
+    } else {
+        _timer.rearm(alarm);
+    }
+}
+
+void backend::wakeup() { _sem.signal(1 - _sem.available_units()); }
+
+[[nodiscard]] ss::future<std::optional<ssx::semaphore_units>> backend::lock() {
+    try {
+        co_return co_await _mutex.get_units(_as);
+    } catch (const ss::broken_semaphore&) {
+        co_return std::nullopt;
+    }
```

Comment:
i think you can use `_mutex.get_units(_as)` directly, as all the places that run with gate already handles the `const ss::broken_semaphore&` exception, so it should be fine to rely on that